### ISSUE TITLE
Does rioxarray allow easy loading of COG overlays into xarray?

### ISSUE THREAD

**Issue by rsignell-usgs** [corteva/rioxarray#45]
Does rioxarray allow loading image data from the COG overlays into xarray?   
I didn't find it but just wanted to make sure I didn't miss it.

**Comment by snowman2** [comment]
`rioxarray` has the `rioxarray.open_rasterio()` function which uses `rasterio.open()` under the hood and should be able to open anything handled by `rasterio.open()`. Do you have an example GOG overlay to try?

**Comment by rsignell-usgs** [comment]
Yes, here's rather lame way I'm dealing with overlays now: https://nbviewer.jupyter.org/gist/rsignell-usgs/0f96bb9c0ca34a5dd0fc8131a7bbae1c
I was hoping for something better.  😺

**Comment by snowman2** [comment]
Nice example. I don't think we have anything better than that at the moment, but it sounds like that would be pretty useful to add.

Currently the open function reads data like so: https://github.com/corteva/rioxarray/blob/83e321c0b3597cc484fb6445d055e4c0ca893f89/rioxarray/_io.py#L132

I think this will need to be thought through a bit as the current implementation uses windowed reading for lazy loading. When using COG overview, it seems that everything needs to be scaled.

My current thoughts on adding support for this is to add support for decimated reading.

Other ideas?

**Comment by rsignell-usgs** [comment]
Before I knew about the existance of rioxarray, I raised the issue here:
https://github.com/pydata/xarray/issues/3269
with an idea at least being able to manually request a specific overlay via an additional parameter request.

**Comment by snowman2** [comment]
Maybe two kwargs?
1. `overlay_index` to select the resampling factor from the overlay index.
2. `resampling_factor` to pass in scale factor https://rasterio.readthedocs.io/en/latest/topics/resampling.html

Also found this useful for reference:
https://automating-gis-processes.github.io/CSC/notebooks/L5/read-cogs.html

**Comment by snowman2** [comment]
https://github.com/mapbox/rasterio/issues/1504

**Comment by snowman2** [comment]
https://gist.github.com/snowman2/9c15df64bd290d1ffd928e0b66855b07
You can try it out with code in #48. It does not allow you to switch between overviews at the moment. So, needs more research. But, the first overview read works.

**Comment by snowman2** [comment]
Issue addressed. It was related to: https://github.com/mapbox/rasterio/issues/1504

**Comment by snowman2** [comment]
@rsignell-usgs any issues using the URL from your example for an example in `rioxarray`? Is it a temporary file or a longer term file?

**Comment by snowman2** [comment]
@vincentsarago do you happen to have a small test-sized COG `tif` file that could be used for testing?

**Comment by vincentsarago** [comment]
@snowman2 https://github.com/cogeotiff/rio-cogeo/blob/master/tests/fixtures/image_rgb.tif ?

**Comment by snowman2** [comment]
![image](https://user-images.githubusercontent.com/8699967/65170498-6f1bc900-da0e-11e9-8183-1e1f2372f275.png)

Do you happen to have one with overviews?

**Comment by snowman2** [comment]
```python
rds.overviews(1)

[]
```

**Comment by vincentsarago** [comment]
Sorry about that, this one should be better https://www.dropbox.com/s/g6etn6o9x3bed62/cog.tif?dl=0

**Comment by snowman2** [comment]
Perfect, thanks! Mind if I add it to the repo? Also, mind if I add you to the contributor list for testing?

**Comment by vincentsarago** [comment]
sure, no worry ;-)

**Comment by rsignell-usgs** [comment]
@snowman2 I intend to leave [the Southern California DEM COG](https://esip-pangeo-uswest2.s3-us-west-2.amazonaws.com/sciencebase/Southern_California_Topobathy_DEM_1m_cog.tif) there for a few months or so, but after that, it will hopefully move back to sciencebase.gov, so probably not the best to use for long-term examples.

**Comment by snowman2** [comment]
> @snowman2 I intend to leave [the Southern California DEM COG](https://esip-pangeo-uswest2.s3-us-west-2.amazonaws.com/sciencebase/Southern_California_Topobathy_DEM_1m_cog.tif) there for a few months or so, but after that, it will hopefully move back to sciencebase.gov, so probably not the best to use for long-term examples.

@rsignell-usgs That is good to know. Thanks for letting me know. Are you aware of any COG files that are accessible via URL and are more long term?

**Comment by vincentsarago** [comment]
@snowman2 on https://openaerialmap.org you should find what you need

**Comment by snowman2** [comment]
> @snowman2 on https://openaerialmap.org you should find what you need

Thanks for the link. I took a quick look and it seems like it provides tiled endpoints, but I couldn't find an endpoint to a COG. I am probably missing something, so please point it out if I am.

**Comment by vincentsarago** [comment]
![Capture d’écran, le 2019-09-20 à 09 57 56](https://user-images.githubusercontent.com/10407788/65332702-2e968980-db8d-11e9-85f0-077277c0872e.png)
@snowman2 you can find the link on the little arrow down

**Comment by snowman2** [comment]
Nice, thanks for that :+1:

**Comment by snowman2** [comment]
Updated: https://corteva.github.io/rioxarray/html/examples/COG.html

**Comment by snowman2** [comment]
@rsignell-usgs, this was a good idea and you provided useful examples to help implement this feature. Mind if I add you to the contributors list for `ideas`?

**Comment by rsignell-usgs** [comment]
@snowman2 , sure that would be great!   BTW, here's my new [streamlined hillshade notebook](https://nbviewer.jupyter.org/gist/rsignell-usgs/f4dd62ad1274c5b5ed69e5a6b81c1295) using the overview enhancement of #48 and the CRS help supplied in #49.

**Comment by snowman2** [comment]
Nice! The notebook looks much simpler to use.

Sounds good, I will add you in a bit.
